p2p/host/basic/basic_host.go

```diff
@@ -343,7 +335,12 @@ func makeUpdatedAddrEvent(prev, current []ma.Multiaddr) *event.EvtLocalAddresses
 	return &evt
 }
 
-func (h *BasicHost) background(p goprocess.Process) {
+func (h *BasicHost) background() {
+	h.refCount.Add(1)
```

Comment:
Unfortunately, this is racy. See the note on https://golang.org/pkg/sync/#WaitGroup.Add.

We have to call Add _before_ we go async.

Reply:
(otherwise `background` might not start running till after the user calls `Close`)

Reply:
Done.

Reply:
I have no idea how I missed this. Thanks for the catch.